Python/Product/Analysis/Interpreter/Ast/AstPythonInterpreterFactory.cs

```diff
@@ -316,17 +319,18 @@ bool useDefaultDatabase
             }
         }
 
-        internal static async Task<IReadOnlyDictionary<string, string>> GetImportableModulesAsync(IEnumerable<string> searchPaths) {
+        internal static async Task<IReadOnlyDictionary<string, string>> GetImportableModulesAsync(IEnumerable<string> searchPaths, CancellationToken cancel) {
             var packageDict = new Dictionary<string, string>();
 
             foreach (var searchPath in searchPaths.MaybeEnumerate()) {
                 IReadOnlyCollection<string> packages = null;
                 if (File.Exists(searchPath)) {
-                    packages = GetPackagesFromZipFile(searchPath);
+                    packages = GetPackagesFromZipFile(searchPath, cancel);
                 } else if (Directory.Exists(searchPath)) {
-                    packages = await Task.Run(() => GetPackagesFromDirectory(searchPath)).ConfigureAwait(false);
+                    packages = await Task.Run(() => GetPackagesFromDirectory(searchPath, cancel)).ConfigureAwait(false);
                 }
                 foreach (var package in packages.MaybeEnumerate()) {
+                    cancel.ThrowIfCancellationRequested();
```

Comment:
Why? `packageDict` is simply a dictionary, adding values to it is very fast. If we made it this far, there is no reason to throw away results that are already acquired.